crates/noirc_evaluator/src/ssa_refactor/ssa_gen/context.rs

```diff
@@ -226,6 +227,22 @@ impl<'a> FunctionContext<'a> {
 
         let mut result = self.builder.insert_binary(lhs, op, rhs);
 
+        let lhs_type = self.builder.current_function.dfg.type_of_value(lhs);
+        let rhs_type = self.builder.current_function.dfg.type_of_value(rhs);
+        if let Some(max_bit_size) =
+            operator_result_max_bit_size_to_truncate(operator, lhs_type, rhs_type)
+        {
+            let result_type = self.builder.current_function.dfg.type_of_value(result);
+            let bit_size = match result_type {
+                Type::Numeric(NumericType::Signed { bit_size })
+                | Type::Numeric(NumericType::Unsigned { bit_size }) => bit_size,
+                _ => {
+                    unreachable!("ICE: Truncation attempted on non-integer");
+                }
+            };
+            result = self.builder.insert_truncate(result, bit_size, max_bit_size);
```

Comment:
I think having a separate `bit_size` and `max_bit_size` is a holdover from the previous IR where `max_bit_size` was an implementation detail to track bit sizes for the removing truncate optimization. If you don't think we need them either, I think we should change the truncate instruction to a more straightforward `v0 = truncate v1 to N bits`

Reply:
Perhaps we could. This information is still required during ACIR gen, but we probably have everything we need to calculate it at the point of use. This involves looking back at the binary instruction being targeted by the `Truncate` during ACIR gen. (We already have to do this anyway for the sake of identifying when to add the integer modulus to prevent subtraction underflow.)

Arguably, you don't need either `bit_size` or `max_bit_size`. `bit_size` is already encoded in the binary instruction's result type. At this point the `Truncate` instruction simply serves as a flag. One might also be tempted to scrap it, and instead make the flag into a property on the binary instruction.

Seems there are a few ways we could rationalise this, which seems out of scope for this PR. I'll open an issue.

Reply:
https://github.com/noir-lang/noir/issues/1604

Reply:
You still need both bit_size and max_bitsize to perform the truncation properly, see: https://github.com/noir-lang/noir/blob/e0ed2913dcc627d721142ab83fdb4a8d747e2351/crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/generated_acir.rs#L104